### ISSUE TITLE
rasterOp: raster area exceedes the dst image

### ISSUE THREAD

**Issue by trufanov-nok** [scantailor/scantailor#210]
I've made a few scans of the book pages - some in 300dpi (text) and most in 600dpi (text and image).
Then I render all of them successfully in 300 dpi to check page order in 3rd party app. I found switching between pages in ST too slow to fast check that no any page is missing.
It was fine so I decided to render all pages in 600dpi and made a djvu from them. Unfortunately ST crashed.
I tried to delete whole cache subfolder and tried to render them again - ST crashed.
I launched ST throw console to get error messages. I've got:
```
Qt has caught an exception thrown from an event handler. Throwing
exceptions from an event handler is not supported in Qt. You must
reimplement QApplication::notify() and catch all exceptions there.

terminate called after throwing an instance of 'std::invalid_argument'
  what():  rasterOp: raster area exceedes the dst image
```

When I'm tracking the progress trhow GUI I'm able to notice that the app crash on one some pages that must be rendered in mixed mode (bw and color modes seems to be fine). Is there a workaround?

**Comment by trufanov-nok** [comment]
It seems it happens in mixed or bw modes (not color/greyscale) for images that content zone has alignment to left or right page border. If I choose center/center-top/center-bottom alignment it renders in 600dpi fine.

**Comment by trufanov-nok** [comment]
I've created a minimalistic project to demonstrate the problem.
https://yadi.sk/d/j33ZjPz6rawXD
There are one image file and one project file. Paths to images were replaced to relative paths in test.ScanTailor file.
Steps to reproduce - open the test project, ignore missing paths, click on latest image on the last step to render it in 600 dpi - ST crashes. ST version is 0.9.11.
Project still has a reference to one missing image that is needed to set up maximal page size for all other pages.

**Comment by trufanov-nok** [comment]
Now when I was going to take a look into this bug by myself I found that my sample files are disappeared. It seems I've accidentally removed them. Let's close this bug till someone could reproduce it or I'll face with such situation one more time.
Moreover, I have assumption where the roots of this problem is, but not sure if it can be fixed elegantly anyway as it seems to be malfunction of heuristic parameter.

**Comment by vphantom** [comment]
I have a mixed mode page that reproduces this. What was your work-around to complete your project?  In my case, going back to "Select Content" and leaving a hair more out from the edge seems to have worked.

**Comment by trufanov-nok** [comment]
@vphantom Could you share your page? You may attach it here.

I did the same - add a thin margin around.

It definitely about margins size. ST adds some "virtual" margin to the image automatically at processing step and then removes it back. Looks like some algorithms expect white margin existence to work. It's fine but the size of this virtual margin is hardcoded and it seems not enough sometime. 
It's fine. What I want is at least to catch this exception and prevent ST from crashing and closing. Ideally would be automatic increase of virtual margin and retry.

**Comment by vphantom** [comment]
@trufanov-nok Argh I'm sorry, I completely forgot to keep a sample page when I finished it not an hour ago. :(  Attached is the smallest 2-page project I could crash ScanTailor with.  It is unfortunately a bit large (5mb), but at least I did use JPEG inside the two TIFF files. [stcrash.zip](https://github.com/scantailor/scantailor/files/872983/stcrash.zip)

I did not know about this virtual margin. Sounds like we might need to enforce a minimum margin greater than zero, although that'd be unfortunate for full-bleed photo books.

(Argh, why doesn't GitHub support tarballs…)

**Comment by trufanov-nok** [comment]
Downloaded, thanks! Will take a look soon.

**Comment by trufanov-nok** [comment]
Could you also attach a *.ScanTailor project file? It could be save at stage 4 just before crash? I can't reproduce crash yet.

**Comment by vphantom** [comment]
The project file is actually included in that ZIP file, in the root called `stcrash.ScanTailor`.  I can't preview the output of either page without crashing with the project in its current state.

**Comment by trufanov-nok** [comment]
Indeed. Now I'm able to reproduce, thanks.

**Comment by trufanov-nok** [comment]
Well, it seems easy to fix.
[Here](https://github.com/scantailor/scantailor/blob/master/filters/output/OutputGenerator.cpp#L248) and on a next (249) line you should replace `toRect()` with `toAlignedRect()` and rebuild. That's all.

The problem must be with rounding rectangles in float coordinates to rectangles in integer coordinates. In some circumstances it rounds to -1 pixel for output rect and to +1 pixel for working rect. Working rect size becomes 1 pixel bigger than output page size and everything collapses. And non-zero margins seems to save ST from that.  `toAlignedRect()` is using `ceil()` instead of `round()` for float coordinates conversion so the sizes should always match.

I still can't say I'm 100% understand the code involved in this so it would be great if you test this fix during your wark and let me know if all your cases are working as expected. Or send me a new page that crashes ST.

**Comment by vphantom** [comment]
Thank you @trufanov-nok for pinpointing this! As I'm already on a private build (for the [normalization bug](https://github.com/scantailor/scantailor/pull/151)) I should be able to test this... right now actually.

Fantastic! I can no longer crash it with my test and I can't see any obvious side-effects. :smiley:

**Comment by trufanov-nok** [comment]
Thanks for testing and reproducible example! Let me know if something goes wrong with this.

**Comment by trufanov-nok** [comment]
Have added one more commit to this patch as I found another example of ST crash with incorrect margins.

**Comment by vphantom** [comment]
Thank you @trufanov-nok!

**Comment by nhertzulis** [comment]
Hi, I'm getting the same error by running

```
scantailor-cli -v -l=1 --enable-page-detection --enable-fine-tuning --enable-auto-margins --content-detection=normal --despeckle=normal --threshold=0 --color-mode=black_and_white --picture-shape=rectangular
```

Could you help me finding a workaround? I'm a little bit lost since it's a bash script done by someone else who isn't available anymore and I don't know bash. Let me know if you need more information

**Comment by trufanov-nok** [comment]
First of all I would note that these parameters:
```--enable-page-detection --enable-fine-tuning --enable-auto-margins --picture-shape=rectangular```
referring to the features from ScanTailor [Enhanced](https://github.com/scantailor/scantailor/tree/enhanced) version. Afaik they still not merged in ST's master branch.
But I don't know if you really need them.

Secondly you seems to forget to apply additional [commit](https://github.com/scantailor/scantailor/pull/270/commits/a97c97112e32e56e8a6230f0b0522535c8b149cc) that referred in this thread.
As I remember even after that I had been approached a few times regarding this bug and made a few more changes to bypass it in my fork. So you may just try to [use it](https://github.com/trufanov-nok/scantailor). It also shall contain Enhanced version features.

**Comment by nhertzulis** [comment]
Yes, I need the enhanced version because there is a bash script that runs different scantailor features according to some input data, the options I gave above are just one example that is already programmed in the script.
Should I build enhanced branch or master from your fork? I tried building master and after installing these packages (among others)

libqt4-dev
libqt4-core
qt4-default
qt4-dev-tools
qt4-linguist-tools

I'm getting this compile error:

CMake Error at CMakeLists.txt:171 (FIND_PACKAGE):
  By not providing "FindQt5Core.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Qt5Core", but
  CMake did not find one.

  Could not find a package configuration file provided by "Qt5Core" with any
  of the following names:

    Qt5CoreConfig.cmake
    qt5core-config.cmake

  Add the installation prefix of "Qt5Core" to CMAKE_PREFIX_PATH or set
  "Qt5Core_DIR" to a directory containing one of the above files.  If
  "Qt5Core" provides a separate development package or SDK, be sure it has
  been installed.

**Comment by trufanov-nok** [comment]
My fork uses Qt 5.x instead of Qt 4.x. So you'll need folowing libs: `qt5-default libtiff5-dev libjpeg-dev libpng-dev libboost-all-dev` Check Example on [this](https://github.com/trufanov-nok/scantailor/wiki/Building-from-Source-Code-on-Linux-and-Mac-OS-X) page. There are also ready binaries including [deb](https://github.com/trufanov-nok/scantailor/releases) package for Linux.

To bould Enhanced version you just ned to get a proper branch from this project:
`git clone -b enhanced https://github.com/trufanov-nok/scantailor.git`
or
```
git clone https://github.com/scantailor/scantailor
cd scantailor
git fetch origin enhanced
git checkout enhanced
```
and then build it with Qt 4.x as described [here](https://github.com/scantailor/scantailor/wiki/Building-from-Source-Code-on-Linux-and-Mac-OS-X)
